lib/basedriver/commands/timeout.js

```diff
@@ -4,23 +4,43 @@ import B from 'bluebird';
 import _ from 'lodash';
 import { util } from 'appium-support';
 import { errors } from '../../mjsonwp';
+import BaseDriver from "../driver";
 
 
 let commands = {}, helpers = {}, extensions = {};
 
 const MIN_TIMEOUT = 0;
 
-commands.timeouts = async function (type, duration) {
-  let ms = this.parseTimeoutArgument(duration);
-  switch (type) {
-    case 'command':
-      this.setNewCommandTimeout(ms);
-      break;
-    case 'implicit':
-      this.setImplicitWait(ms);
-      break;
-    default:
-      throw new Error(`Invalid timeout '${type}'`);
+// If we define `commands.timeouts` instead of `commands.timeoutsW3C`, the command `timeouts` will be called
+// from other dirver's timeouts. See https://github.com/appium/appium-base-driver/pull/164
+// Arguments will be: [{"protocol":"W3C","implicit":30000}, "1dcfe021-8fc8-49bd-8dac-e986d3091b97", ...]
+// eslint-disable-next-line no-unused-vars
+commands.timeouts = async function (timeoutsObj) {
+  if (timeoutsObj.protocol === BaseDriver.DRIVER_PROTOCOL.W3C) {
+    const {script, pageLoad, implicit} = timeoutsObj;
+    log.debug(`script: ${script}, pageLoad: ${pageLoad}, implicit: ${implicit}`);
+
+    if (util.hasValue(script) || util.hasValue(pageLoad)) {
+      throw new errors.NotImplementedError('Not implemented yet for script and pageLoad.');
+    }
+
+    if (util.hasValue(implicit)) {
+      this.implicitWait(implicit);
+    }
+  } else {
+    const {type, ms} = timeoutsObj;
+    log.debug(`type: ${type}, ms: ${ms}`);
+
+    switch (type) {
+      case 'command':
+        this.newCommandTimeout(ms);
+        break;
+      case 'implicit':
+        this.implicitWait(ms);
+        break;
+      default:
+        throw new Error(`Invalid timeout '${type}'`);
```

Comment:
What happens for drivers that have more timeouts than this?

Reply:
I assume it's overridden in such drivers or you'd prefer to simply ignore unknown values instead of throwing an error?

Reply:
If sub-classed drivers are meant to override this function, then all the logic for handling W3C vs JSONWP is either lost or needs to be rewritten there.

As I suggested before, this `timeouts` function should do something like
```js
if (timeoutsObj.protocol === BaseDriver.DRIVER_PROTOCOL.W3C) {
  await this.setW3CTimeouts(timeoutsObj);
} else {
  const {type, ms} = timeoutsObj;
  await this.setJSONWPTimeouts(type, ms);
}
```
And then individual drivers should override the `setW3CTimeouts` and `setJSONWPTimeouts` functions.